### PR TITLE
Add enable_auto_compaction option

### LINKED ISSUE #9186: Need a way to disable auto-compaction on startup
Until we have https://github.com/scylladb/scylla/issues/5489
there are cases where we want scylla to not start regular compaction on startup,
for example, when running validation compaction, that, at the moment, does not
quarantine sstables that failed validation.
If regular compaction runs in the background it races with the admin and may very well compact away
those sstables and spread the suspected corruption to other sstables.

My proposal is to allow configuring compaction_static_shares to zero and disable auto-compaction in this case.
auto-compaction can then be re-enabled later using `nodetool enableautocompaction`.

Note that we currently don't support dynamically updating compaction_static_shares when the configuration is reloaded.
Once we do, it'd be helpful also to enableautocompaction if it was disabled previously due to `compaction_static_shares == 0`.

### PR DESCRIPTION
**Pull request by bhalevy**
Support configuring automatic compaction on startup.
    
Log a global message when auto-compaction is disabled.
    
Auto-compaction can be reenabled using `nodetool enableautocompaction`
or dynamically when reloading the configuration.
    
The motivation for providing this option is to disable
auto-compaction, e.g. before running scrub/validation compaction
since sstables that fail validation race with regular compaction
that might compact them away and spread the possible corruption
to new sstables before the admin has a chance to handle them.

Fixes #9186

### CODE REVIEW

**Review by denesb — COMMENTED** [review]
I wonder if instead of adding a specific option for this, it would be worth adding a more general "service-mode" option, under which we could continue to collect related changes, slowly working towards https://github.com/scylladb/scylla/issues/5489.

**Comment by bhalevy** [comment]
> I wonder if instead of adding a specific option for this, it would be worth adding a more general "service-mode" option, under which we could continue to collect related changes, slowly working towards #5489.

But wouldn't it be misleading until it does enough to be considered as "service mode"?

**Comment by denesb** [comment]
> > I wonder if instead of adding a specific option for this, it would be worth adding a more general "service-mode" option, under which we could continue to collect related changes, slowly working towards #5489.
> 
> But wouldn't it be misleading until it does enough to be considered as "service mode"?

Not if we document what exactly it does. I don't see this option being used during regular operation, and if the only use-case is that of "service-mode" we better add a `service_mode` option instead.

**Comment by bhalevy** [comment]
I'm not sure.
I can see use cases in dtest that would disable auto compaction on startup and do stuff when the node is operational.

@roy, @amoskong @bentsi, what do you think?

**Comment by denesb** [comment]
> I'm not sure.
> I can see use cases in dtest that would disable auto compaction on startup and do stuff when the node is operational.
> 
> @roy, @amoskong @bentsi, what do you think?

If there are standalone use-cases then sure, we can make this a flag of its own. Doesn't the existing REST API already cover the QA use case?

**Comment by bhalevy** [comment]
> > I'm not sure.
> > I can see use cases in dtest that would disable auto compaction on startup and do stuff when the node is operational.
> > @roy, @amoskong @bentsi, what do you think?
> 
> If there are standalone use-cases then sure, we can make this a flag of its own. Doesn't the existing REST API already cover the QA use case?

Unfortunately not all use cases are covered since until the api is started and `nodetool disableautocompaction` is run scylla already starts to auto-compact both system and user tables.

**Comment by denesb** [comment]
Yes, the question is whether any test really wants to disable auto-compaction for startup.

**Comment by bhalevy** [comment]
For the validation tests I created the tables with `NullCompactionStrategy` to achieve this effect.

File: table.cc [hunk 1]
```diff
@@ -2115,7 +2114,7 @@ void
 table::enable_auto_compaction() {
     // XXX: unmute backlog. turn table backlog back on.
     //      see table::disable_auto_compaction() notes.
-    _compaction_disabled_by_user = false;
+    _config.enable_auto_compaction = true;
```
Comment (denesb) [inline table.cc:2117]: Won't this affect all tables now, as `_config.enable_auto_compaction` is a live value.

Comment (bhalevy) [reply table.cc:2117]: yes, you're right.
we need per table state in this case.

Comment (denesb) [reply table.cc:2117]: I think we can just retain the existing field in `table` to act as a local override, and we `&&` the global and local config to compute the value relevant for the table.

File: main.cc [hunk 2]
```diff
@@ -1076,12 +1076,16 @@ int main(int ac, char** av) {
                 }
             }
 
-            db.invoke_on_all([] (database& db) {
+            if (cfg->enable_auto_compaction()) {
+              db.invoke_on_all([] (database& db) {
                 for (auto& x : db.get_column_families()) {
                     table& t = *(x.second);
                     t.enable_auto_compaction();
```
[the remark this answers is not in the document]
Comment (bhalevy) [reply main.cc:1083]: It is disabled by default in distributed_loader::populate_column_family:
https://github.com/scylladb/scylla/blob/6dd67012bbc3ada84d66f0ee436b408a820528b9/distributed_loader.cc#L559

[the remark this answers is not in the document]
Comment (denesb) [reply main.cc:1083]: Ok.

**Comment by bhalevy** [comment]
retracting

**Pull request closed without merging**